testing/src/main/java/org/oppia/testing/TestCoroutineDispatchers.kt

```diff
@@ -0,0 +1,116 @@
+
+  /**
+   * Runs all current tasks pending, but does not follow up with executing any tasks that are
+   * scheduled after this method finishes.
+   *
+   * Note that it's generally not recommended to use this method since it may result in
+   * unanticipated dependencies on the order in which this class processes tasks for each handled
+   * thread and coroutine dispatcher.
+   */
+  @ExperimentalCoroutinesApi
+  fun runCurrent() {
+    do {
+      flushNextTasks()
+    } while (hasPendingCompletableTasks())
+  }
+
+  /**
+   * Advances the system clock by the specified time in milliseconds and then ensures any new tasks
+   * that were scheduled are fully executed before proceeding. This does not guarantee the
+   * dispatchers enter an idle state, but it should guarantee that any tasks previously not executed
+   * due to it not yet being the time for them to be executed may now run if the clock was
+   * sufficiently forwarded.
+   */
+  @ExperimentalCoroutinesApi
+  fun advanceTimeBy(delayTimeMillis: Long) {
+    fakeSystemClock.advanceTime(delayTimeMillis)
+    runCurrent()
+  }
+
+  /**
+   * Runs all tasks on all tracked threads & coroutine dispatchers until no other tasks are pending.
+   * However, tasks that require the clock to be advanced will likely not be run (depending on
+   * whether the test under question is using a paused execution model, which is recommended for
+   * Robolectric tests).
+   */
+  @ExperimentalCoroutinesApi
+  fun advanceUntilIdle() {
+    do {
+      flushAllTasks()
+    } while (hasPendingTasks())
+  }
+
+  @ExperimentalCoroutinesApi
+  private fun flushNextTasks() {
+    if (backgroundTestDispatcher.hasPendingCompletableTasks()) {
+      backgroundTestDispatcher.runCurrent()
+    }
+    if (blockingTestDispatcher.hasPendingCompletableTasks()) {
+      blockingTestDispatcher.runCurrent()
+    }
+    if (!shadowUiLooper.isIdle) {
+      shadowUiLooper.idle()
+    }
+  }
+
+  @ExperimentalCoroutinesApi
+  private fun flushAllTasks() {
+    val currentTimeMillis = fakeSystemClock.getTimeMillis()
+    if (backgroundTestDispatcher.hasPendingCompletableTasks()) {
+      backgroundTestDispatcher.advanceUntilIdle()
+    }
+    if (blockingTestDispatcher.hasPendingCompletableTasks()) {
+      blockingTestDispatcher.advanceUntilIdle()
+    }
+    shadowUiLooper.idleFor(currentTimeMillis, TimeUnit.MILLISECONDS)
+    SystemClock.setCurrentTimeMillis(currentTimeMillis)
+  }
+
+  private fun hasPendingTasks(): Boolean {
+    // TODO(#89): Ensure the check for pending UI thread tasks is actually correct.
+    return backgroundTestDispatcher.hasPendingTasks()
+        || blockingTestDispatcher.hasPendingTasks()
+        || !shadowUiLooper.isIdle
+  }
+
+  private fun hasPendingCompletableTasks(): Boolean {
+    return backgroundTestDispatcher.hasPendingCompletableTasks()
+        || blockingTestDispatcher.hasPendingCompletableTasks()
+        || !shadowUiLooper.isIdle
```

Comment:
Indentation Update:
```
  private fun hasPendingTasks(): Boolean {
    // TODO(#89): Ensure the check for pending UI thread tasks is actually correct.
    return backgroundTestDispatcher.hasPendingTasks()
      || blockingTestDispatcher.hasPendingTasks()
      || !shadowUiLooper.isIdle
  }

  private fun hasPendingCompletableTasks(): Boolean {
    return backgroundTestDispatcher.hasPendingCompletableTasks()
      || blockingTestDispatcher.hasPendingCompletableTasks()
      || !shadowUiLooper.isIdle
  }
```

Or are we keeping continuation indentation as `4 spaces` ?

Reply:
Looks like we're using 4 spaces for continuation, but I apparently incorrectly put ``||`` at the beginning of the line. I used ktlint to reformat the file as part of answering your question. I think we should follow the formatter in this case unless other folks have a strong opinion about it.

Reply:
Okay, now it looks more readable. I think we should use ktlint only to avoid any confusion.